tardis/plasma/properties/partition_function.py

```diff
@@ -51,6 +50,260 @@ def calculate(levels, beta_rad, w):
+            r_ul_index = lnl * number_of_levels + lnu
+
+            r_ul_matrix = np.zeros((number_of_levels, number_of_levels,
+                len(t_electron)), dtype=np.float64)
+            r_ul_matrix_reshaped = r_ul_matrix.reshape((number_of_levels**2,
+                len(t_electron)))
+            r_ul_matrix_reshaped[r_ul_index] = A_uls[np.newaxis].T + \
+                B_uls[np.newaxis].T * j_blues.ix[j_blues_index]
+            r_ul_matrix_reshaped[r_ul_index] *= beta_sobolevs[lines_index]
+
+            r_lu_matrix = np.zeros_like(r_ul_matrix)
+            r_lu_matrix_reshaped = r_lu_matrix.reshape((number_of_levels**2,
+                len(t_electron)))
+            r_lu_matrix_reshaped[r_lu_index] = B_lus[np.newaxis].T * \
+                j_blues.ix[j_blues_index] * beta_sobolevs[lines_index]
+
+            if atomic_data.has_collision_data:
+                if previous_electron_densities is None:
+                    collision_matrix = r_ul_matrix.copy()
+                    collision_matrix.fill(0.0)
+                else:
+                    collision_matrix = nlte_data.get_collision_matrix(species,
+                        t_electron) * previous_electron_densities.values
+            else:
+                collision_matrix = r_ul_matrix.copy()
+                collision_matrix.fill(0.0)
+
+            rates_matrix = r_lu_matrix + r_ul_matrix + collision_matrix
+
+            for i in xrange(number_of_levels):
+                rates_matrix[i, i] = -rates_matrix[:, i].sum(axis=0)
+
+            rates_matrix[0, :, :] = 1.0
+
+            x = np.zeros(rates_matrix.shape[0])
+            x[0] = 1.0
+            for i in xrange(len(t_electron)):
+                level_boltzmann_factor = \
+                    np.linalg.solve(rates_matrix[:, :, i], x)
+                general_level_boltzmann_factor[i].ix[species] = \
+                    level_boltzmann_factor
+        return general_level_boltzmann_factor
+'''
+class LevelBoltzmannFactorNLTE(ProcessingPlasmaProperty):
+    outputs = ('level_boltzmann_factor',)
+
+    def __init__(self, plasma_parent, classical_nebular=False,
+        coronal_approximation=False):
+        super(LevelBoltzmannFactorNLTE, self).__init__(plasma_parent)
+        if classical_nebular == True:
+            self.calculate = self._calculate_classical_nebular
+        elif coronal_approximation == True:
+            self.calculate = self._calculate_coronal_approximation
+        else:
+            self.calculate = self._calculate_general
+        self._update_inputs()
+
+    def _calculate_classical_nebular(self, general_level_boltzmann_factor):
+        return general_level_boltzmann_factor
+    def _calculate_coronal_approximation(self, general_level_boltzmann_factor):
+        return general_level_boltzmann_factor
+    def _calculate_general(self, general_level_boltzmann_factor):
+        return general_level_boltzmann_factor
+    def _main_nlte_calculation(self):
+        pass
+'''
```

Comment:
Hi @wkerzendorf. Just trying to get the basic structure here to work like you suggested earlier. When I run this I get the error, 'Can't instantiate abstract class LevelBoltzmannFactorNLTE with abstract methods calculate.' I assume that's just because there's no explicit `def calculate()`? How do I make this work?